PWG/EMCAL/EMCALtasks/AliEmcalCorrectionClusterNonLinearityMCAfterburner.cxx

```diff
@@ -0,0 +1,565 @@
+{
+}
+
+/**
+ * Initialize and configure the component.
+ */
+Bool_t AliEmcalCorrectionClusterNonLinearityMCAfterburner::Initialize()
+{
+  // Initialization
+  AliEmcalCorrectionComponent::Initialize();
+
+  std::string nonLinMethodStr = "";
+  GetProperty("afterburnerVersion", nonLinMethodStr);
+  EMCAfterburnerMethod_t fAfterburnerMethod = fgkMCNonlinearityAfterburnerMethodMap.at(nonLinMethodStr);
+  GetProperty("setMCPeriod", fMCPeriod);
+  GetProperty("setForceClusterE", fSetForceClusterE);
+
+  //Load the right period string
+  TString period = SummarizeMCProductions(fMCPeriod);
+  //Load the right parameters for the MC afterburner correction
+  InitNonLinearityParam(period, fAfterburnerMethod);
+
+  return kTRUE;
+}
+
+/**
+ * Create run-independent objects for output. Called before running over events.
+ */
+void AliEmcalCorrectionClusterNonLinearityMCAfterburner::UserCreateOutputObjects()
+{   
+  AliEmcalCorrectionComponent::UserCreateOutputObjects();
+  
+  // Create my user objects.
+  if (fCreateHisto){
+    fEnergyDistBefore = new TH1F("hEnergyDistBefore","hEnergyDistBefore;E_{clus} (GeV)",1500,0,150);
+    fOutput->Add(fEnergyDistBefore);
+    fEnergyTimeHistBefore = new TH2F("hEnergyTimeDistBefore","hEnergyTimeDistBefore;E_{clus} (GeV);time (s)",1500,0,150,500,-1e-6,1e-6);
+    fOutput->Add(fEnergyTimeHistBefore);
+    fEnergyDistAfter = new TH1F("hEnergyDistAfter","hEnergyDistAfter;E_{clus} (GeV)",1500,0,150);
+    fOutput->Add(fEnergyDistAfter);
+    fEnergyTimeHistAfter = new TH2F("hEnergyTimeDistAfter","hEnergyTimeDistAfter;E_{clus} (GeV);time (s)",1500,0,150,500,-1e-6,1e-6);
+    fOutput->Add(fEnergyTimeHistAfter);
+    
+    // Take ownership of output list
+    fOutput->SetOwner(kTRUE);
+  }
+}
+
+/**
+ * Called for each event to process the event data.
+ */
+Bool_t AliEmcalCorrectionClusterNonLinearityMCAfterburner::Run()
+{
+  AliEmcalCorrectionComponent::Run();
+
+  // loop over clusters
+  AliVCluster *clus = 0;
+  AliTLorentzVector clusVec;
+  AliClusterContainer * clusCont = 0;
+  TIter nextClusCont(&fClusterCollArray);
+
+  //Apply the nonlinearity afterburner correction for the MC.
+  //Parameters are set in InitNonLinearityParam() and copied from AliCaloPhotonCuts.cxx
+  //Apply the MC nonlinearity functions with the set parameters
+  if (fNonLinearityAfterburnerFunction != 0) {
+
+    while ((clusCont = static_cast<AliClusterContainer*>(nextClusCont()))) {
+
+      if (!clusCont) return kFALSE;
+      auto clusItCont = clusCont->all_momentum();
+
+      for (AliClusterIterableMomentumContainer::iterator clusIterator = clusItCont.begin(); clusIterator != clusItCont.end(); ++clusIterator) {
+        clus    = static_cast<AliVCluster *>(clusIterator->second);
+        clusVec = static_cast<AliTLorentzVector>(clusIterator->first);
+
+        if (!clus->IsEMCAL()) continue;
+        //currently only for EMCal clusters. Skip DCal clusters
+        AliFiducialCut* fiducialCuts = new AliFiducialCut();
```

Comment:
I missed this one before, but this is leaking memory! Please remember that every `new` should be matched with a corresponding `delete`. But there is also no reason that I see to use a pointer and to create a new object every loop. Please move this definition up before the for loop and just define it as `AliFiducialCuts fiducialCuts;`

Reply:
If the object is used exclusively in the function / loop then please do not allocate the memory with new at all but rather put the object on the stack

```
AliFiducialCut fiducialCuts;
fiducialCuts.IsInFiducialCut(...);
````

Reply:
As suggested above, you could just change this to directly check the phi value of the cluster, and get rid of all this AliFiducial cut business.

Reply:
changed to the phi value suggestion of James! Much better